Rework how the project directory is determined for jack-in and connect

PEZ opened the pull request:
First we try find the workspace root, _then_ we add custom sub dir

Hello, can you review this fix of `getProjectDir()` for me? It is not supposed to add any fancy new possibilities, just fix so that a custom project root (sub directory) should work in more cases. It is a very central function so I'd appreciate an extra brain on it. Cheers!

PEZ committed:
Apply custom project root last

First we try find the workspace root, _then_ we add custom sub dir

PEZ commented:
Please merge it if you think it is good enough. :smile:

kstehn reviewed (commented):
Looks good.
But i think i found a problem , and it might be the reason why i thought this is buggy.

When you Jack-In the Terminal says "Executing task in folder <firstFolder> ...."
After that i always canceled the process. This time i just let it run and it is the right Project.
But i have no idea where that might happen

PEZ committed:
More helpful error message for project root dir

PEZ committed:
Use Task() constructor w/o folder

PEZ commented:
I think I might have been using the wrong constructor overload for the Task all this time. Fixed that now. Can you try and see if it fixes the problem? If so we should have a much more reliable jack-in for projects not in the workspace root now.

PEZ committed:
Find project dir through current file

Removes `projectRootDir` setting and uses this logic instead:

1. If there is no file open. Stop and complain.
2. If there is a file open, use it to determine the project root
by looking for project files from the file's directory and up to
the window root (for plain folder windows) or the file's
workspace folder root (for workspaces) to find the project root.
Return
1. the window root for plain folders
2. first workspace root for workspaces.

PEZ committed:
Reomve projectRootDirectory setting lookup

PEZ committed:
Expect to find nrepl port files ni project root

kstehn commented on calva/connector.ts:
```diff
@@ -400,22 +400,13 @@ export let cljSession: NReplSession;
-    } else {
-        return path.join(util.getProjectDir(), subPath);
+    try {
+        const projectRoot = util.getProjectDir(),
```
Small error here  it should be a` ;` not` ,`

kstehn commented:
Also like we discussed to remind you, add some documentation to the function that you are currently working on :).

PEZ committed:
Use worskapce.findFiles

PEZ committed:
First initialize the project root, then use it

PEZ commented:
Now I have made the project root something we need to initialize as the start of the jack-in and (standalone) connect processes. When the path is actually needed, this initialized value will be used. Should be both more efficient and less error prone.

I also added that doc string, finally. :smile: Which was a very good thing to add, because the way I had phrased the strategy used made me realize that this new way of finding the project files isn't implementing that strategy! This will now fail for the ”plain” folder, multi-project case. (I haven't tested that yet, but I am pretty certain.) Committing this anyway, so that I can get your help with sanity checking the general change for init-then-use.

Will restore the old new way of finding the project files now. Then, maybe, this is closing in on being ready to merge.

PEZ committed:
Look for project files from opened file and up

PEZ committed:
Fix broken shadow-cljs jack-in

PEZ committed:
Fix broken clojure-cli jack-in

PEZ commented:
This now works for the plain folder multi project case. I've tested it with Leiningen, Clojure CLI and shadow-cljs. An inconvenience is that Calva remembers the jack-in and connect picker selections per workspace which can be confusing and trick people to accept the wrong defaults. I think we should fix that before shipping. What say you?

I'll make some tests with a real workspace now. Please put it to a test on your side as well.

PEZ committed:
Make vscode report the correct jack-in WS folder

PEZ committed:
Don't feed getWorkspaceFolders with null

PEZ committed:
typo

PEZ committed:
Save picker defaults per project root

PEZ commented:
VSIX package built and sent to `calva-dev` in Clojurians Slack. Let's give it a day or two. If no one reports a problem, then merge! : smile:

mrkam2 commented:
Hi @PEZ I found that broke my Jack-In presets. I need "projectRootDir" setting so that I can choose _another_ project having the same source file. Is it possible to restore it?

mrkam2 commented:
Sorry, I got it wrong. I was actually using a custom Calva build from https://github.com/BetterThanTomorrow/calva/pull/580.

PEZ commented:
So you're good now?

mrkam2 commented:
> So you're good now?

Thanks for checking.

I need that PR to be merged (as I need projectRootDir setting). I have a temporary workaround for now.
